### ISSUE TITLE
bug(vscode) - `command / ctrl -Enter` rewrites recent changes...

### ISSUE THREAD

**Issue by jduckles** [stencila/stencila#2388]
Version: v0.0.6

When appending to a code block, and then hitting `command-Enter` the most recent change is deleted:

https://github.com/user-attachments/assets/40916623-a111-4c5d-b568-540aed5b6a85

Based on a conversation with @nokome he thought it might be due to a 500ms bounce. I waited longer than 500ms and still see the issue. I can duplicate the 500ms bounce issue, and that one is annoying for someone who touch types `command-Enter` almost subconsciously.

Digging a bit more, an unclosed raw fence above is maybe causing this... 

https://github.com/user-attachments/assets/8dd6acb5-552d-44d9-8ef8-38917d9c90c1

The dangling fence is removed (unexpected) and changes to the code block aren't saved. Also unexpected. I had this happening in a document repeatedly, but I can't reproduce that now. Fortunately Command-Z returns the missing text, but still the kind of weirdness that will drive user away.

**Comment by nokome** [comment]
Thanks for digging into this.

When there are unbalanced backtick or colons fences  the document is treated as invalid, and no changes to it are merged into the in-memory node tree. This is necessary to maintain the state of nodes (e.g. execution outputs, provenance). If we didn't do that then if you entered three or more backticks midway through a document all the nodes below that would be treated as code and all their state lost.

The Stencila Language Server emits diagnostics for unbalanced fences and ignores updates until issues are resolved and the document is in a valid state. However, it should also refuse to run executable nodes (or any code lens / command probably) because the the `code` property of those nodes will be out of sync with what is in the text document.

As an aside, the loss of code observed here (updating the document with an out of sync state) should be fixed by df8cd096006b23a45ac84a310c4b3c28eb58cf17.

**Comment by jduckles** [comment]
Just a small update that this has continued to bite me now and again.

**Comment by nokome** [comment]
Yes, me too 😊 . I've added a `sync_state` to the LSP document's so we should now be able to fix this by adding the following to the `execute_command` function.

```rs
    // Wait for the document to be synced
    sync_state_receiver
        .wait_for(|sync_state| matches!(sync_state, SyncState::Updated))
        .await
        .map_err(|_| {
            ResponseError::new(ErrorCode::INTERNAL_ERROR, "Unable to wait for is_synced")
        })?;
```

**Comment by nokome** [comment]
This should now be fixed by:

https://github.com/stencila/stencila/blob/f5a8d0bd7accbc460d68e3c2afdd61d545c0a9a8/rust/lsp/src/commands.rs#L146-L171
